packages/firestore-types/index.d.ts

```diff
@@ -181,7 +192,11 @@ export class FirebaseFirestore {
    * @param documentPath A slash-separated path to a document.
    * @return The `DocumentReference` instance.
    */
-  doc(documentPath: string): DocumentReference;
+  doc(documentPath: string): DocumentReference<DocumentData>;
+  doc<T>(
```

Comment:
I don't think this should be typed. You can have a typed collection that has a typed DocumentReferences, but if you call `doc()`  from an untyped Collection or FirebaseFirestore, my gut feeling is that it is better not to allow types.